File: plover_vm/macro/isa.py
"""v0.1 macro ISA opcodes."""

from __future__ import annotations

OP_ADD = 0x01
OP_LDA = 0x02
OP_STA = 0x03
OP_BEQ = 0x04
OP_JMP = 0x05
OP_CALL = 0x06
OP_RET = 0x07
OP_LDIO = 0x08
OP_STIO = 0x09
OP_HALT = 0x0A
OP_ADD_RR = 0x0B  # R2 <- R0 + R1
OP_MOV = 0x0C  # deprecated on breadboard; VM compat — use TFR opcodes
OP_CMP = 0x0D  # R0 - imm, set flags
OP_BCS = 0x0E  # branch if carry (unsigned >=)
OP_STA16 = 0x0F  # store R0 to abs16 (3-byte insn)

# Implied register transfer (v1.0 breadboard — 1 byte, no operand)
OP_TFR01 = 0x10  # R0 <- R1
OP_TFR02 = 0x11  # R0 <- R2
OP_TFR10 = 0x12  # R1 <- R0
OP_TFR12 = 0x13  # R1 <- R2
OP_TFR20 = 0x14  # R2 <- R0
OP_TFR21 = 0x15  # R2 <- R1

TFR_OPS = frozenset({OP_TFR01, OP_TFR02, OP_TFR10, OP_TFR12, OP_TFR20, OP_TFR21})

# VM-only 16-bit wide register ops (not on breadboard ISA)
OP_WADD_RR = 0x20  # W2 <- W0 + W1
OP_WMOV = 0x21  # imm: (dst<<4)|src
OP_WCMP16 = 0x22  # W0 vs imm16 unsigned

WIDE_IMM16_OPS = frozenset({OP_WCMP16})

# dst, src for TFR opcodes (R0=0, R1=1, R2=2)
TFR_REG_MAP: dict[int, tuple[int, int]] = {
    OP_TFR01: (0, 1),
    OP_TFR02: (0, 2),
    OP_TFR10: (1, 0),
    OP_TFR12: (1, 2),
    OP_TFR20: (2, 0),
    OP_TFR21: (2, 1),
}

PHASE_COUNTS: dict[int, int] = {
    OP_ADD: 3,
    OP_LDA: 2,
    OP_STA: 2,
    OP_BEQ: 2,
    OP_JMP: 1,
    OP_CALL: 1,
    OP_RET: 1,
    OP_LDIO: 2,
    OP_STIO: 2,
    OP_HALT: 1,
    OP_MOV: 1,
    OP_CMP: 3,
    OP_STA16: 2,
    OP_TFR01: 1,
    OP_TFR02: 1,
    OP_TFR10: 1,
    OP_TFR12: 1,
    OP_TFR20: 1,
    OP_TFR21: 1,
}

WIDE_ABS16_OPS = frozenset({OP_BEQ, OP_JMP, OP_CALL, OP_STA16})


def phase_count(opcode: int) -> int:
    return PHASE_COUNTS.get(opcode & 0xFF, 1)

File: crates/plover_core/src/isa.rs
pub const OP_ADD: u8 = 0x01;
pub const OP_LDA: u8 = 0x02;
pub const OP_STA: u8 = 0x03;
pub const OP_BEQ: u8 = 0x04;
pub const OP_JMP: u8 = 0x05;
pub const OP_CALL: u8 = 0x06;
pub const OP_RET: u8 = 0x07;
pub const OP_LDIO: u8 = 0x08;
pub const OP_STIO: u8 = 0x09;
pub const OP_HALT: u8 = 0x0A;
pub const OP_ADD_RR: u8 = 0x0B;
pub const OP_MOV: u8 = 0x0C;
pub const OP_CMP: u8 = 0x0D;
pub const OP_BCS: u8 = 0x0E;
pub const OP_STA16: u8 = 0x0F;
pub const OP_TFR01: u8 = 0x10;
pub const OP_TFR02: u8 = 0x11;
pub const OP_TFR10: u8 = 0x12;
pub const OP_TFR12: u8 = 0x13;
pub const OP_TFR20: u8 = 0x14;
pub const OP_TFR21: u8 = 0x15;
pub const OP_WADD_RR: u8 = 0x20;
pub const OP_WMOV: u8 = 0x21;
pub const OP_WCMP16: u8 = 0x22;

pub const TFR_OPS: [u8; 6] = [
    OP_TFR01, OP_TFR02, OP_TFR10, OP_TFR12, OP_TFR20, OP_TFR21,
];

pub const WIDE_IMM16_OPS: [u8; 1] = [OP_WCMP16];
pub const WIDE_ABS16_OPS: [u8; 4] = [OP_BEQ, OP_JMP, OP_CALL, OP_STA16];

pub fn is_tfr(op: u8) -> bool {
    TFR_OPS.contains(&op)
}

pub fn tfr_regs(op: u8) -> Option<(usize, usize)> {
    match op {
        OP_TFR01 => Some((0, 1)),
        OP_TFR02 => Some((0, 2)),
        OP_TFR10 => Some((1, 0)),
        OP_TFR12 => Some((1, 2)),
        OP_TFR20 => Some((2, 0)),
        OP_TFR21 => Some((2, 1)),
        _ => None,
    }
}

pub fn is_wide_imm16(op: u8) -> bool {
    WIDE_IMM16_OPS.contains(&op)
}

pub fn is_wide_abs16(op: u8) -> bool {
    WIDE_ABS16_OPS.contains(&op)
}

pub fn phase_count(opcode: u8) -> usize {
    match opcode {
        OP_ADD => 3,
        OP_LDA | OP_STA | OP_BEQ | OP_LDIO | OP_STIO | OP_STA16 => 2,
        OP_CMP => 3,
        OP_JMP | OP_CALL | OP_RET | OP_HALT | OP_MOV => 1,
        OP_TFR01 | OP_TFR02 | OP_TFR10 | OP_TFR12 | OP_TFR20 | OP_TFR21 => 1,
        _ => 1,
    }
}
